### PR TITLE
Ability to add new routes to the routes map array, in a each zend module.

### PR DESCRIPTION
**Pull request by yurige-matrix**
Hi @bradymiller . In this commit, we added the ability to extend the map of routes for "REST API" requests. We are planning to begin with development of client-side app and this functionality is necessary for our modules. What do you think of it ? Have a nice day!.

### CODE REVIEW

**Comment by bradymiller** [comment]
hi @adunsulag , Hoping you have time to review this PR? thanks, -brady

File: _rest_routes.inc.php [hunk 1]
```diff
@@ -328,6 +329,10 @@
 
 );
 
+$restApiExtendEvent = new RestApiExtendEvent(RestConfig::$ROUTE_MAP);
+$restApiExtendEvent = $GLOBALS["kernel"]->getEventDispatcher()->dispatch(RestApiExtendEvent::EVENT_HANDLE, $restApiExtendEvent, 10);
```
Comment (adunsulag) [inline _rest_routes.inc.php]: Do you plan on only extending the REST api, or will you be extending the FHIR api also?  @sjpadgett do you have any thoughts on this?

You could leverage the same event for both the REST and FIHR pieces, but you would need to pass in some context to say which one is being extended...

If you don't care about FIHR we can treat it as a separate event.

Comment (sjpadgett) [reply _rest_routes.inc.php]: Does it matter? Which api used is determined by the route.

Comment (adunsulag) [reply _rest_routes.inc.php]: @sjpadgett you're right that the api used is determined by the route, but this pull request is attempting to modify just the routes in $ROUTE_MAP and not the ones in $FHIR_ROUTE_MAP.

Comment (sjpadgett) [reply _rest_routes.inc.php]: duh Jerry! Right, Yes we'd want FHIR inclusion. I am wondering now if I should have just designed for one route array. I was thinking it more efficient memory wise to separate them.

File: tests/eventdispatcher/RestApiEventHookExample/Module.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    /*
+     * A function that adds new routes to a route map array
+     * instead some_section need use a real name of section name
+     * */
+    public function addRestAPIRouteToMap($m){
+        $extend_api = [
+                        "GET /api/some_route" => function (){
+                            RestConfig::authorization_check("some_section", "users");
+                            return (new SomeClassWithSomeLogics())->getAll();
```
[the remark this answers is not in the document]
Comment (yurige-matrix) [reply tests/eventdispatcher/RestApiEventHookExample/Module.php]: In general, I removed an unnecessary class. I return an array with three digits

File: _rest_routes.inc.php [hunk 3]
```diff
@@ -328,6 +329,10 @@
 
 );
 
+$restApiExtendEvent = new RestApiExtendEvent(RestConfig::$ROUTE_MAP);
```
[the remark this answers is not in the document]
Comment (yurige-matrix) [reply _rest_routes.inc.php]: This is a great idea!

File: src/Events/RestApiExtend/RestApiExtendEvent.php [hunk 4]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+namespace OpenEMR\Events\RestApiExtend;
+
+use Symfony\Component\EventDispatcher\Event;
+
+class RestApiExtendEvent extends Event
+{
+
+    const EVENT_HANDLE = 'restConfig.route_map.extend';
```
[the remark this answers is not in the document]
Comment (yurige-matrix) [reply src/Events/RestApiExtend/RestApiExtendEvent.php]: This is also a great idea!

File: _rest_routes.inc.php (hunk 1 above)
Comment (yurige-matrix) [reply _rest_routes.inc.php]: In the next commit, I will try to implement this feature.

**Comment by adunsulag** [comment]
@bradymiller This looks good to me, if you or @sjpadgett had any additional feedback feel free to jump in.  Otherwise I'm fine with taking it in.

**Comment by sjpadgett** [comment]
Thanks @adunsulag I'm bringing this in just so I can feel important today:)
@yurige-matrix So glad to see the api being used for client app.  I know of at least 6 apps by others in the wild, two of which are FHIR. It would be nice if you could share what Matrix new client app will be doing.

**Pull request merged**